### ISSUE TITLE
Refactor pub-sub / client events

### ISSUE THREAD

**Issue by anbraten** [woodpecker-ci/woodpecker#742]
We currently have a "interesting" pub-sub mechanism in woodpecker which works, but could get some love. I would like to refactor it a bit to have a more generic event system. For example at the moment if a build-step was updated, the event contains the complete repo, build and build step data, which seems quite heavy to me. Instead I would like some kind of `{ type: 'pipeline-step', data: ...  }` structure approach even if this would require use to send two events where we currently used one event like `{ repo: ...., build: ..., proc: ... }`

**Comment by 6543** [comment]
yes I'm thinking of clustering and ha for quite a while, and the two things holding us back are: how to cluster gRPC & pub-sub

**Comment by anbraten** [comment]
Clustering pub-sub should easily be possible by writing a redis / rabbitmq implementation of [this interface](https://github.com/woodpecker-ci/woodpecker/blob/master/server/pubsub/pubsub.go). gRPC can probably simply be clustered by putting a load-balancer in front of it. The queue would be an additional part which could need some kind of clustering, but may be covered by a centeral database already.

**Comment by 6543** [comment]
I would like to have an implementation of pub-sub based on redis & also for the queue (if we still need it after refactoring ...) to be able to use redis

but it should be opt-in!

also we additionaly could look at [NATS](https://nats.io/) ...

**Comment by anbraten** [comment]
> maybe add some mq and split server to server(user/ui/api) and control, control can be singleton, but api-server can be multiple servers

But this would be a single point of failure again. In addition I would currently prefer to stick to a single server binary as it makes the deployment much simpler (think of RaspberryPIs etc users).
